internal/types/subtypes/interceptor.go

```diff
@@ -0,0 +1,193 @@
+package subtypes
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/sdk/pbs/controller/protooptions"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+func messageDomain(m proto.Message) string {
+	r := m.ProtoReflect()
+	fd := r.Descriptor().ParentFile()
+	if fd == nil {
+		return ""
+	}
+
+	opts, ok := fd.Options().(*descriptorpb.FileOptions)
+	if !ok {
+		return ""
+	}
+
+	domain := proto.GetExtension(opts, protooptions.E_Domain).(string)
+	return domain
+}
+
+func transformRequestAttributes(_ context.Context, req proto.Message) proto.Message {
+	domain := messageDomain(req)
+
+	r := req.ProtoReflect()
+	fields := r.Descriptor().Fields()
+
+	itemField := fields.ByName("item")
+	idField := fields.ByName("id")
+	attributesField := fields.ByName("attributes")
```

Comment:
Could we add a comment here to explain the assumptions that underlay these? E.g.

```
* "item", if it exists, is assumed to be a field in the top level request containing the primary model of the service, e.g. an Account used in a Create request.
* "id", if it exists, is assumed to be a field in the model containing the models id, from which we can infer the subtype of the request.
* "attributes" is assumed to be a field in the model containing a `google.protobuf.Struct` with fields mapping to one of the specific subtypes, or a generic attribute structure if the type is not specific.
```

Reply:
I added docs in [7658e9e](https://github.com/hashicorp/boundary/pull/1967/commits/7658e9e489454006241899d5864127cb99e47422) that should hopefully explain this now